### ISSUE TITLE
Simple throwing kernel hangs

### ISSUE THREAD

**Issue by tgymnich** [JuliaGPU/Metal.jl#433]
This simple kernel will cause a hang leading to 100% GPU usage and taking down the WindowServer eventually. A similar Metal/C++ kernel runs just fine. The generated IR is also mostly identical.

Another interesting observation is that one can save the WindowServer from its death spiral by running the good C++ kernel.

```julia
using Metal

function kernel()
    throw(nothing)
    return nothing
end

Metal.@sync @metal threads=8 kernel()
```

```c++
void gpu_report_exception() {
    return;
}

void gpu_signal_exception() {
    return;
}

kernel void add_arrays()
{
    gpu_report_exception();
    gpu_signal_exception();

    __builtin_trap();
    __builtin_unreachable();
}
```

**Comment by tgymnich** [comment]
Taking the metallib that worked in the xcode example and running it with Metal.jl also hangs!
The same also holds in reverse: Running the metallib generated by Metal.jl with the Xcode example does not hang.

**Comment by tgymnich** [comment]
This does not hang with ObjectiveC.jl:

```julia
using ObjectiveC, .Foundation

@objcwrapper MTLDevice <: NSObject
dev = MTLDevice(ccall(:MTLCreateSystemDefaultDevice, id{MTLDevice}, ()))


@objcwrapper immutable=false MTLLibrary <: NSObject
err = Ref{id{NSError}}(nil)
file = NSString("/Users/tim/Desktop/default.metallib")
handle = @objc [dev::id{MTLDevice} newLibraryWithFile:file::id{NSString} error:err::Ptr{id{NSError}}]::id{MTLLibrary}
lib = MTLLibrary(handle)


@objcwrapper immutable=false MTLFunction <: NSObject
name = NSString("add_arrays")
handle = @objc [lib::id{MTLLibrary} newFunctionWithName:name::id{NSString}]::id{MTLFunction}
fun = MTLFunction(handle)


@objcwrapper immutable=false MTLComputePipelineState <: NSObject
err = Ref{id{NSError}}(nil)
handle = @objc [dev::id{MTLDevice} newComputePipelineStateWithFunction:fun::id{MTLFunction} error:err::Ptr{id{NSError}}]::id{MTLComputePipelineState}
pip = MTLComputePipelineState(handle)


@objcwrapper MTLCommandQueue <: NSObject
handle = @objc [dev::id{MTLDevice} newCommandQueue]::id{MTLCommandQueue}
queue = MTLCommandQueue(handle)


@objcwrapper MTLCommandBuffer <: NSObject
@objcproperties MTLCommandBuffer begin
    @autoproperty status::NSUInteger
    @autoproperty error::id{NSError}
end

handle = @objc [queue::id{MTLCommandQueue} commandBuffer]::id{MTLCommandBuffer}
cmdbuf = MTLCommandBuffer(handle)


@objcwrapper MTLComputeCommandEncoder <: NSObject
handle = @objc [cmdbuf::id{MTLCommandBuffer} computeCommandEncoder]::id{MTLComputeCommandEncoder}
cce = MTLComputeCommandEncoder(handle)


struct MTLSize
    width::NSUInteger
    height::NSUInteger
    depth::NSUInteger

    MTLSize(w=1, h=1, d=1) = new(w, h, d)
end

threadGroupSize = MTLSize(1,0,0)
gridSize = MTLSize(1,0,0)


@objc [cce::id{MTLComputeCommandEncoder} setComputePipelineState:pip::id{MTLComputePipelineState}]::Nothing
@objc [cce::id{MTLComputeCommandEncoder} dispatchThreadgroups:gridSize::MTLSize threadsPerThreadgroup:threadGroupSize::MTLSize]::Nothing
@objc [cce::id{MTLComputeCommandEncoder} endEncoding]::Nothing



@objc [cmdbuf::id{MTLCommandBuffer} commit]::Nothing
@objc [cmdbuf::id{MTLCommandBuffer} waitUntilCompleted]::Nothing

cmdbuf.status
```

**Comment by maleadt** [comment]
Doesn't reproduce on latest master with macOS 26.

**Comment by christiangnrd** [comment]
I can still reproduce this with macOS 26.5 from the REPL. I also encountered the WindowServer issues mentioned in the issue description

**Comment by maleadt** [comment]
Hmm, thanks for confirming. I'm working on porting the non-blocking synchronization from CUDA.jl, so maybe that does anything.

**Comment by maleadt** [comment]
@christiangnrd Can you test https://github.com/JuliaGPU/Metal.jl/pull/783? I still can't manage to get this to hang here.

**Comment by christiangnrd** [comment]
Still hangs :/

**Comment by maleadt** [comment]
Ugh. I'm not sure what's up now. Here's some suggestions by Claude to further debug:

> To figure out whether the GPU is actually wedged or it's a host-side issue, run this script and paste the output. It launches the throwing kernel and then submits a trivial "probe" command buffer on the same queue — command buffers run in order, so the probe's status tells us whether the GPU ever finishes the trapping kernel.
>  
> ```julia
> using Metal
> using Metal: MTL
> 
> function kernel()
>     throw(nothing)
>     return nothing
> end
> 
> queue = Metal.global_queue(Metal.device())
> @metal threads=8 kernel()          # launch, do NOT synchronize
> 
> probe = MTL.MTLCommandBuffer(queue) # trivial buffer queued right after
> MTL.commit!(probe)
> 
> names = Dict(0=>"NotEnqueued", 1=>"Enqueued", 2=>"Committed",
>              3=>"Scheduled", 4=>"Completed", 5=>"Error")
> for i in 1:50
>     s = Int(probe.status)
>     Core.println("poll $i: probe status = $s ($(get(names, s, \"?\")))")
>     s >= 4 && (Core.println("probe finished — GPU not wedged"); break)
>     sleep(0.2)
> end
> ```
> 
> What the output means:
> - Probe reaches 4 (Completed) or 5 (Error) → the GPU finishes fine and the hang is host-side. On this machine I get Completed on the first poll.
> - Probe stays stuck at 2 (Committed) / 3 (Scheduled) for all 50 polls → the trapping kernel is spinning your GPU and never completing. (Check Activity Monitor → GPU; it'll be pegged near 100%.)
> 
> If it stays stuck, please also grab a native backtrace of the hung process so we can see exactly where it's parked. In a second terminal while a hang is in progress (e.g. run the original Metal.@sync @metal
> threads=8 kernel() in one REPL):
> 
> ```
> # find the Julia process id:
> pgrep -fl julia
> # sample it for 5s (replace <PID>):
> sample <PID> 5 -file /tmp/metal_hang.txt
> ```
> 
> and attach /tmp/metal_hang.txt.
> 
> Finally, please post:
> - your GPU/chip — output of Metal.device()
> - your macOS build — sw_vers
> - whether Ctrl-C interrupts the hung Metal.@sync REPL (it should, on the latest version), even if the GPU stays busy.

**Comment by christiangnrd** [comment]
The probe shows scheduled (3) for all checks and then I had to hard reset my laptop even though it appeared to have cleanly exited

**Comment by maleadt** [comment]
I can reproduce on an M1, without a way to unblock the GPU (so needing a reboot).

**Comment by maleadt** [comment]
I _can_ reproduce this with pure ObjectiveC and a metallib containing just a trap, so this isn't Metal.jl's launch code after all.

[metal433-repro.zip](https://github.com/user-attachments/files/28301883/metal433-repro.zip)

**Comment by maleadt** [comment]
@christiangnrd Can you test with https://github.com/JuliaGPU/GPUCompiler.jl/pull/810?

**Comment by christiangnrd** [comment]
Metal.jl main branch and JuliaGPU/GPUCompiler.jl#810 doesn't hang.

**Comment by maleadt** [comment]
Ah finally, thanks for confirming. That will finally allow us to unblock https://github.com/JuliaGPU/Metal.jl/pull/418
